File: OC_多种cell的处理优化/Cell/ViewController.m
//
//  ViewController.m
//  Cell
//
//  Created by suifumin on 2020/11/11.
//

#import "ViewController.h"
#import "AModel.h"
#import "BModel.h"
#import "CModel.h"
#import "DModel.h"
#import "BaseCell.h"
@interface ViewController ()<UITableViewDelegate,UITableViewDataSource>
@property (nonatomic,strong) UITableView *tableView;
@property(nonatomic,strong)NSMutableArray * dataArray;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    AModel *modelA = [[AModel alloc]init];
    BModel *bmodel = [[BModel alloc]init];
    CModel *cmodel = [[CModel alloc]init];
    DModel *dmodel = [[DModel alloc]init];
    self.dataArray = [NSMutableArray array];
    [self.dataArray addObject:modelA];
    [self.dataArray addObject:bmodel];
    [self.dataArray addObject:cmodel];
    [self.dataArray addObject:dmodel];
    
    self.tableView = [[UITableView alloc]initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    self.tableView.backgroundColor = [UIColor whiteColor];
    [self.view addSubview:self.tableView];
    self.tableView.tableFooterView = [[UIView alloc]init];
    [self.tableView reloadData];
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataArray.count;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    id <ModelConfigProtocol>model = self.dataArray[indexPath.row];
    BaseCell *cell = (BaseCell *)[model tableView:tableView reuseIdenyifier:[model cellReuseIdentifier]];
    cell.model = model;
    return cell;
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    id <ModelConfigProtocol>model = self.dataArray[indexPath.row];
    return [model cellHeightWithindexPath:indexPath];
}
@end

File: swift_多种cell的处理/Cell/ViewController.swift
//
//  ViewController.swift
//  Cell
//
//  Created by suifumin on 2020/11/12.
//

import UIKit

class ViewController: UIViewController {
    lazy var tableView: UITableView = {
        let tableView = UITableView(frame: .zero, style: .plain)
        tableView.backgroundColor = .white
        tableView.tableFooterView = UIView()
        tableView.dataSource = self
        tableView.delegate = self
        return tableView
    }()
    lazy var dataArray:[ModelProtocol] = [ModelProtocol]()
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.addSubview(self.tableView)
        self.tableView.frame = self.view.bounds
        let amodel = Amodel()
        let bmodel = Bmodel()
        let cmodel = Cmodel()
        self.dataArray.append(amodel)
        self.dataArray.append(bmodel)
        self.dataArray.append(cmodel)
        self.tableView.reloadData()
        
            
        
    }


}
extension ViewController: UITableViewDataSource,UITableViewDelegate {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        self.dataArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let model: ModelProtocol = self.dataArray[indexPath.row]
        let cell: UITableViewCell = model.createCellWithTabelView(tableView: tableView, reusedStr: model.cellReusedIdentifier())
        return cell
    }
    
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let model: ModelProtocol = self.dataArray[indexPath.row]
        return model.cellHeigtForIndexth(indexpath: indexPath)
    }
    
    
    
    
    
    
    
}
